### PR TITLE
Parser error handling

### PR DESCRIPTION
**Pull request by kpodsiad**
Closes https://github.com/VirtusLab/scala-yaml/issues/12

### CODE REVIEW

File: src/main/scala/org/virtuslab/internal/load/parse/ParserImpl.scala [hunk 1]
```diff
@@ -39,68 +39,96 @@ private enum Production:
...
+          else Right(acc :+ event)
+        case Left(err) => Left(err)
     }
-    Right(loop(Nil, List(ParseStreamStart)))
+    loop(Nil, List(ParseStreamStart))
 
-  def getNextEvent(in: Reader, stack: List[Production]): (Event, List[Production]) = {
+  type EventResult = Either[YamlError, (Event, List[Production])]
+
+  private def error(expected: String, got: Token): EventResult = Left(
```
Comment (lwronski) [inline src/main/scala/org/virtuslab/internal/load/parse/ParserImpl.scala:58]: We should separate the errors logic from the parsing. We should cover error parsing with new type error. 
In new type error, you can create helper method for parsing, such as 

`EventError.from(expected: Token, got: Token): EventResult`

**Review by lwronski — COMMENTED** [review]
We should separate errors logic from parsing.

File: src/main/scala/org/virtuslab/internal/load/parse/ParserImpl.scala (hunk 1 above)
Comment (kpodsiad) [reply src/main/scala/org/virtuslab/internal/load/parse/ParserImpl.scala:58]: Should we change `YamlError` to sealed trait and then, `EventError` will extend it?

Comment (lwronski) [reply src/main/scala/org/virtuslab/internal/load/parse/ParserImpl.scala:58]: Yes, good idea

**Review by lwronski — APPROVED** [review]
LGTM

**Pull request merged**